### DISCUSSION TITLE
Gift Card information on Cart Object

### DISCUSSION THREAD

**Discussion by igoraugustynski** [swellstores/community#285] [Help] [answered]
Hello,
I have a couple questions regarding gift cards and swell-js.
Let's say I have a cart to which I added 3 different gift cards. On the cart object returned by swell-js, I get:
```
giftcards: [
    {
        "code": "94CGJWUYJK7ZYKYH",
        "id": "64908a92076b5400128499f9",
        "codeFormatted": "94CG JWUY JK7Z YKYH",
        "amount": 65,
        "last4": "YKYH"
    },
    {
        "code": "WTKT6RF3QG7V9MBF",
        "id": "6499c79316cd280013b55af2",
        "codeFormatted": "WTKT 6RF3 QG7V 9MBF",
        "amount": 2,
        "last4": "9MBF"
    },
    {
        "code": "C3KHAEV6BVXZUCSZ",
        "id": "6499cc007fb2bf0012c5755f",
        "codeFormatted": "C3KH AEV6 BVXZ UCSZ",
        "amount": 100,
        "last4": "UCSZ"
    }
]
```

Here are my questions:
- How do I know the amount that is going to be withdrawn from each card?
- Is there a way to get more information about the card, i.e. the original value? 

Also, in this case, `giftcardTotal` is 167 even if the order total is just 70$. Is this correct?

**Comment by ericingram** [comment]
@igoraugustynski the `giftcards.amount` field specifies how much will be spent from each card, in order. You can change this value if you want. It defaults to the balance on the card, but again you can change the value if you want to control how much is able to be spent on each card. The current balance is available via backend API but probably not the frontend API, besides this property.

When the order is submitted, the system will walk through each card and accumulate the order total. So for example, if the order total is $50, then $50 would be subtracted from the first card and no more from the other cards.

Hope that helps!

**Reply by igoraugustynski** [reply]
Well this does not track at all. In the example I pasted above, `cart.grandTotal` was 70$, but as You can see the sum of `giftcards.amount` is 167$. It seems that `giftcards.amount` is actually the current balance of the card, or there is a bug in how it's calculated. The same applies for `cart.giftcardTotal` - It looks like it represents the sum of all current balances of the cards applied to the cart.

**Reply by ericingram** [reply]
> It defaults to the balance on the card

You’re correct

**Reply by ericingram** [reply]
I should have said, it specifies how much *can* be spent on the card, meaning it’s the card balance

**Reply by igoraugustynski** [reply]
@ericingram I have a couple more question about Gift Cards. Let's suppose that a user made a purchase for a total of 80 euro with a Gift Card which originally had a 100 euro balance - so there is still 20 euro left on the card. Is it automatically linked to the user account if an order is made with it? Is there a way, maybe using the FE API, to show the gift cards and their remaining balances to the user? I can't seem to find anything about it in the docs.

**Reply by ericingram** [reply]
There are options for assigning a gift card balance to a user account, or not, leaving it to the gift card itself which can then be used by another account.

@logeshswell can probably expand on the different approaches

**Reply by igoraugustynski** [reply]
How do I assign the Gift Card to a user when applying it to the cart?

**Reply by logeshswell** [reply]
> there is still 20 euro left on the card. Is it automatically linked to the user account if an order is made with it? 

The remaining balance will still remain in the gift card for future payments. You also have the option to move it to the customer account balance by passing `account_id` field in `/giftcards` endpoint using backend API.

**Reply by igoraugustynski** [reply]
So the balance does not get automatically transferred to the user account when used. Is there a way to do this using the FE API?

**Reply by logeshswell** [reply]
In FE API, you should be able to do it if the permissions are allowed for `/giftcards` endpoint.
```ts
const giftcard = await swell.put('/giftcards/<giftcard_id>', {
  account_id: '<account_id>'
});
```

API permissions can be allowed as follows:
<img width="507" alt="Screenshot 2023-08-28 at 5 40 15 PM" src="https://github.com/swellstores/community/assets/100766131/bbbb9558-c6e5-47f6-aa00-5432bd27429d">

Documentation reference:
https://developers.swell.is/guides/opening-content-permissions

**Reply by logeshswell** [reply, marked as answer]
> Is there a way, maybe using the FE API, to show the gift cards and their remaining balances to the user?

There is no straight-forward method, but you can follow the below steps to solve your use case.

1. In **Giftcards** model, add a custom field **Order Account** of type **Lookup** to **Customers** model.

<img width="558" alt="Screenshot 2023-08-28 at 6 40 00 PM" src="https://github.com/swellstores/community/assets/100766131/d8b0db79-3f9b-4e5a-9c9f-1ccb8dc569f8">

2. Set the custom field as read-only and formula to store the value automatically in POST/PUT API requests.

<img width="510" alt="Screenshot 2023-08-28 at 6 24 49 PM" src="https://github.com/swellstores/community/assets/100766131/b36c1483-6567-43c5-9937-0487e91006e7">

3. Open up read permissions for giftcards model. You can include as many fields as required to show up in the frontend.

<img width="509" alt="Screenshot 2023-08-28 at 6 52 35 PM" src="https://github.com/swellstores/community/assets/100766131/f50d9d95-57f8-4554-a913-ffca54862369">

4. Fetch the list of giftcards purchased by the current user.

```ts
const account = await swell.account.get();
const giftcards = await swell.get('/giftcards', {
  order_account_id: account.id
});
```
